Python/ceval_macros.h

```diff
@@ -271,8 +306,9 @@ GETITEM(PyObject *v, Py_ssize_t i) {
 /* This takes a uint16_t instead of a _Py_BackoffCounter,
  * because it is used directly on the cache entry in generated code,
  * which is always an integral type. */
+// Force re-specialization when tracing a side exit to get good side exits.
 #define ADAPTIVE_COUNTER_TRIGGERS(COUNTER) \
-    backoff_counter_triggers(forge_backoff_counter((COUNTER)))
+    backoff_counter_triggers(forge_backoff_counter((COUNTER))) || IS_JIT_TRACING_MAKING_PROGRESS()
```

Comment:
```suggestion
    backoff_counter_triggers(forge_backoff_counter((COUNTER)))
```

You can guarantee progress, but not switch from non-tracing to tracing mode at except at the end of an instruction.

Reply:
If we don't have the `|| IS_JIT_TRACING()`. How do we force the JIT to re-specialize when tracing? Recall that the specializing interpreter is perfectly fine executing the base op. However, we don't want that, we want to actively force it to re-specialize in the case of a deopt/miss. This contributed to around 30% of the speedup in richards, as now we handle side exits/deopts perfectly.